### PR TITLE
Goreleaser

### PR DESCRIPTION
**Pull request by spinillos**
Adds `Goreleaser` to generate semantic versioning to thema.

Contributes to: https://github.com/grafana/thema/issues/172

### CODE REVIEW

File: .goreleaser.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      - linux
+      - windows
+      - darwin
+    binary: thema
+    main: ./cmd/thema
+
+archives:
+  - format: tar.gz
+    format_overrides:
+    - goos: windows
```
Comment (joanlopez) [inline .goreleaser.yaml:17]: For Windows, we'll also probably need to add `.exe` extension to the binary file.

**Review by joanlopez — COMMENTED** [review]
Thanks so much @spinillos! 
Looks like a good first iteration, I've just left few comments I think we could iterate a bit before shipping it!

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+.PHONY: gen-release
+
+gen-release:
+	./scripts/check_semantic_versioning.sh $(tag)
+	git tag $(tag)
+	git push origin $(tag)
+	goreleaser --rm-dist
```
[the remark this answers is not in the document]
Comment (spinillos) [reply Makefile:7]: Yes, we can do it. When we create the tag should be enough, right?

File: .goreleaser.yaml (hunk 1 above)
Comment (spinillos) [reply .goreleaser.yaml:17]: I didn't add anything because I wasn't really sure. This is from the basic template.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (joanlopez) [reply Makefile:7]: I think so, yeah! At least for now that should be enough, later we can change it to something different if needed, but changing "the trigger" (every tag, on every commit.. whatever) shouldn't be a big deal later.

File: .goreleaser.yaml (hunk 1 above)
Comment (joanlopez) [reply .goreleaser.yaml:17]: Sure, np! I experienced that in a different project, that's why I'm suggesting it hahaha, but my first experience was exactly as yours (following the basic template), so don't worry at all!

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (sdboyer) [reply Makefile:7]: agreed, let's automate based on a tag existing

File: .goreleaser.yaml (hunk 1 above)
Comment (spinillos) [reply .goreleaser.yaml:17]: I'm seeing into `goreleaser` documentation that available formats are the following:
```
Valid options are `tar.gz`, `tgz`, `tar.xz`, `txz`, tar`, `gz`, `zip` and `binary`.
```

So I understand that could we use `binary`?

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (spinillos) [reply Makefile]: I added into `go.yml` workflow but I need to set `GITHUB_TOKEN` if we don't have it. @sdboyer do you have permissions to set that variable into settings?

[the remark this answers is not in the document]
Comment (sdboyer) [reply Makefile]: i've made @grafana/grafana-as-code maintainers - does that let you do it?

[the remark this answers is not in the document]
Comment (spinillos) [reply Makefile]: I have access to settings now but looks like that add env variables needs a higher permission level 😅

[the remark this answers is not in the document]
Comment (AgnesToulet) [reply Makefile]: This env variable should already exist (at least, this worked when I did that for the TF schemas provider release, see https://github.com/grafana/terraform-provider-schemas/blob/main/.github/workflows/release.yml#L40).

[the remark this answers is not in the document]
Comment (spinillos) [reply Makefile]: Thanks @AgnesToulet!!

File: .goreleaser.yaml (hunk 1 above)
Comment (joanlopez) [reply .goreleaser.yaml:17]: I think the current format (`zip`) is fine, because it will generate a compressed file with the resulting binary inside (similar to other platforms, just replacing tar by zip, which is way more common on Windows) - that's ok!

What I suggested at the beginning is changing the name of the resulting binary (what will end up within that zip file), from just `thema` (or whatever it is now) to `thema.exe`, because I think that, based on previous experiences, on Windows (by default), a file needs to have the `.exe` extension to be considered an executable 😅 

Hope it makes more sense now! 🙏🏻

File: docs/release_new_version.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Release a new version
+This process uses [GoReleaser](https://goreleaser.com/install/) to generate the release.
+
+1. Execute `make gen-release tag=your_tag` using [semantic versioning](https://goreleaser.com/limitations/semver/).
```
Comment (joanlopez) [inline docs/release_new_version.md:4]: Looks like this will require an update after the recent changes 😛

**Comment by sdboyer** [comment]
thanks!

**Comment by sdboyer** [comment]
Hmm, there seems to be an issue with the tag regex: https://github.com/grafana/thema/actions/runs/5713950449. Annoying that the actions engine only complained _after_ merging...

But, it's not immediately obvious to me why that's a problem, as the regex itself [is valid](https://regex101.com/r/Fq8VdI/2). @spinillos could you investigate and fix?

**Comment by joanlopez** [comment]
> Hmm, there seems to be an issue with the tag regex: https://github.com/grafana/thema/actions/runs/5713950449. Annoying that the actions engine only complained _after_ merging...
> 
> But, it's not immediately obvious to me why that's a problem, as the regex itself [is valid](https://regex101.com/r/Fq8VdI/2). @spinillos could you investigate and fix?

Related with: https://github.com/orgs/community/discussions/55854
TL;DR: Looks like it has no full regex capabilities.

So, I guess we'll need to either simplify it or find an alternative.
After a very quick search, I've found a [Parse Semver](https://github.com/marketplace/actions/parse-semver) action, which I guess could help to define a conditional workflow.
